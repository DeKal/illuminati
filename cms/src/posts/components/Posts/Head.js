import React from 'react'
import PropTypes from 'prop-types'
import { styled } from '@material-ui/core/styles'
import TableCell from '@material-ui/core/TableCell'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import TableSortLabel from '@material-ui/core/TableSortLabel'
import Checkbox from '@material-ui/core/Checkbox'

const EnhancedTableHead = ({
  order,
  orderColumn,
  numSelected,
  rowCount,
  setOrder,
  setOrderColumn,
  selectAllPosts,
  clearSelected
}) => {
  const createSortHandler = property => () => {
    const isAsc = orderColumn === property && order === 'asc'
    setOrder(isAsc ? 'desc' : 'asc')
    setOrderColumn(property)
  }

  return (
    <TableHead>
      <TableRow>
        <TableCell padding="checkbox">
          <Checkbox
            indeterminate={numSelected > 0 && numSelected < rowCount}
            checked={rowCount > 0 && numSelected === rowCount}
            onChange={event => {
              if (event.target.checked) {
                selectAllPosts()
                return
              }
              clearSelected()
            }}
            inputProps={{ 'aria-label': 'select all desserts' }}
          />
        </TableCell>
        {headCells.map(headCell => (
          <TableCell
            key={headCell.id}
            align={headCell.numeric ? 'right' : 'left'}
            padding={headCell.disablePadding ? 'none' : 'default'}
            sortDirection={orderColumn === headCell.id ? order : false}
          >
            <TableSortLabel
              active={orderColumn === headCell.id}
              direction={orderColumn === headCell.id ? order : 'asc'}
              onClick={createSortHandler(headCell.id)}
            >
              {headCell.label}
              {orderColumn === headCell.id ? (
                <VisuallyHiddenSpan>
                  {order === 'desc' ? 'sorted descending' : 'sorted ascending'}
                </VisuallyHiddenSpan>
              ) : null}
            </TableSortLabel>
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  )
}

EnhancedTableHead.propTypes = {
  numSelected: PropTypes.number,
  order: PropTypes.oneOf(['asc', 'desc']).isRequired,
  orderColumn: PropTypes.string.isRequired,
  rowCount: PropTypes.number.isRequired,
  setOrder: PropTypes.func,
  setOrderColumn: PropTypes.func,
  selectAllPosts: PropTypes.func,
  clearSelected: PropTypes.func
}

export default EnhancedTableHead

const headCells = [
  {
    id: 'title',
    numeric: false,
    disablePadding: true,
    label: 'Title'
  },
  { id: 'author', numeric: false, disablePadding: false, label: 'Author' },
  { id: 'tag', numeric: false, disablePadding: false, label: 'Tags' },
  { id: 'commentNum', numeric: true, disablePadding: false, label: 'Comments' },
  { id: 'date', numeric: false, disablePadding: false, label: 'Date' }
]

const VisuallyHiddenSpan = styled('span')({
  border: 0,
  clip: 'rect(0 0 0 0)',
  height: 1,
  margin: -1,
  overflow: 'hidden',
  padding: 0,
  position: 'absolute',
  top: 20,
  width: 1
})
