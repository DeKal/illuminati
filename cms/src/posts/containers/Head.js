import { connect } from 'react-redux'
import Head from 'posts/components/Posts/Head'
import {
  setOrder,
  setOrderColumn,
  selectAllPosts,
  clearSelected
} from 'posts/state/actions'
import { getOrder, getOrderColumn } from 'posts/state/selectors'

const mapStateToProps = state => ({
  order: getOrder(state),
  orderColumn: getOrderColumn(state)
})

const mapDispatchToProps = {
  setOrder,
  setOrderColumn,
  selectAllPosts,
  clearSelected
}

export default connect(mapStateToProps, mapDispatchToProps)(Head)
