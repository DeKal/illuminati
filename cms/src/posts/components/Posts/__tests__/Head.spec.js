import * as React from 'react'
import { shallow } from 'enzyme'
import Head from '../Head'

describe('Head', () => {
  it('should render with no error', function() {
    const wrap = shallow(
      <Head order={'asc'} orderColumn={'title'} numSelected={0} rowCount={5} />
    )
    expect(wrap).toMatchSnapshot()
  })

  it('should render Head with indetermine state checkbox', function() {
    const wrap = shallow(
      <Head order={'asc'} orderColumn={'title'} numSelected={2} rowCount={5} />
    )
    expect(wrap).toMatchSnapshot()
  })
  it('should render Head with all check checkbox', function() {
    const wrap = shallow(
      <Head order={'asc'} orderColumn={'title'} numSelected={5} rowCount={5} />
    )
    expect(wrap).toMatchSnapshot()
  })

  it('should handle sort column asc', function() {
    const setOrder = jest.fn()
    const setOrderColumn = jest.fn()
    const wrap = shallow(
      <Head
        order={'asc'}
        orderColumn={'title'}
        numSelected={0}
        rowCount={5}
        setOrder={setOrder}
        setOrderColumn={setOrderColumn}
      />
    )
    wrap.find('[data-test-id="head-sort-title"]').simulate('click')
    expect(setOrder.mock.calls.length).toBe(1)
    expect(setOrder.mock.calls[0][0]).toEqual('desc')
    expect(setOrderColumn.mock.calls.length).toBe(1)
    expect(setOrderColumn.mock.calls[0][0]).toBe('title')
  })

  it('should handle sort desc an asc column', function() {
    const setOrder = jest.fn()
    const setOrderColumn = jest.fn()
    const wrap = shallow(
      <Head
        order={'asc'}
        orderColumn={'title'}
        numSelected={0}
        rowCount={5}
        setOrder={setOrder}
        setOrderColumn={setOrderColumn}
      />
    )
    wrap.find('[data-test-id="head-sort-author"]').simulate('click')
    expect(setOrder.mock.calls.length).toBe(1)
    expect(setOrder.mock.calls[0][0]).toEqual('asc')
    expect(setOrderColumn.mock.calls.length).toBe(1)
    expect(setOrderColumn.mock.calls[0][0]).toBe('author')
  })
})
