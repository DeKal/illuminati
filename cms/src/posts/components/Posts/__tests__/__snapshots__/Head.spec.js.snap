// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Head should render Head with all check checkbox 1`] = `
<WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableCell))
      padding="checkbox"
    >
      <WithStyles(ForwardRef(Checkbox))
        checked={true}
        indeterminate={false}
        inputProps={
          Object {
            "aria-label": "select all posts",
          }
        }
        onChange={[Function]}
      />
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="title"
      padding="none"
      sortDirection="asc"
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={true}
        data-test-id="head-sort-title"
        direction="asc"
        onClick={[Function]}
      >
        Title
        <Styled(span)>
          sorted ascending
        </Styled(span)>
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="author"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-author"
        direction="asc"
        onClick={[Function]}
      >
        Author
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="date"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-date"
        direction="asc"
        onClick={[Function]}
      >
        Date
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="tag"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-tag"
        direction="asc"
        onClick={[Function]}
      >
        Tags
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="right"
      key="commentNum"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-commentNum"
        direction="asc"
        onClick={[Function]}
      >
        Comments
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
  </WithStyles(ForwardRef(TableRow))>
</WithStyles(ForwardRef(TableHead))>
`;

exports[`Head should render Head with indetermine state checkbox 1`] = `
<WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableCell))
      padding="checkbox"
    >
      <WithStyles(ForwardRef(Checkbox))
        checked={false}
        indeterminate={true}
        inputProps={
          Object {
            "aria-label": "select all posts",
          }
        }
        onChange={[Function]}
      />
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="title"
      padding="none"
      sortDirection="asc"
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={true}
        data-test-id="head-sort-title"
        direction="asc"
        onClick={[Function]}
      >
        Title
        <Styled(span)>
          sorted ascending
        </Styled(span)>
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="author"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-author"
        direction="asc"
        onClick={[Function]}
      >
        Author
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="date"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-date"
        direction="asc"
        onClick={[Function]}
      >
        Date
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="tag"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-tag"
        direction="asc"
        onClick={[Function]}
      >
        Tags
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="right"
      key="commentNum"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-commentNum"
        direction="asc"
        onClick={[Function]}
      >
        Comments
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
  </WithStyles(ForwardRef(TableRow))>
</WithStyles(ForwardRef(TableHead))>
`;

exports[`Head should render with no error 1`] = `
<WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableCell))
      padding="checkbox"
    >
      <WithStyles(ForwardRef(Checkbox))
        checked={false}
        indeterminate={false}
        inputProps={
          Object {
            "aria-label": "select all posts",
          }
        }
        onChange={[Function]}
      />
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="title"
      padding="none"
      sortDirection="asc"
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={true}
        data-test-id="head-sort-title"
        direction="asc"
        onClick={[Function]}
      >
        Title
        <Styled(span)>
          sorted ascending
        </Styled(span)>
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="author"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-author"
        direction="asc"
        onClick={[Function]}
      >
        Author
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="date"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-date"
        direction="asc"
        onClick={[Function]}
      >
        Date
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="left"
      key="tag"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-tag"
        direction="asc"
        onClick={[Function]}
      >
        Tags
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell))
      align="right"
      key="commentNum"
      padding="default"
      sortDirection={false}
    >
      <WithStyles(ForwardRef(TableSortLabel))
        active={false}
        data-test-id="head-sort-commentNum"
        direction="asc"
        onClick={[Function]}
      >
        Comments
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
  </WithStyles(ForwardRef(TableRow))>
</WithStyles(ForwardRef(TableHead))>
`;
